### ISSUE TITLE
Allow pattern search in cwd or allow fully qualified executables

### ISSUE THREAD

**Issue by armaxri** [matepek/vscode-catch2-test-adapter#444]
**Checklist**

- [x] Checked [Support page](https://github.com/matepek/vscode-catch2-test-adapter/blob/master/documents/support.md)

**Is your feature request related to a problem? Please describe.**
We have extremely large workspaces with quite a few test executables, which have to be listed a few times in the extension due to different execution parameters. This causes even very powerful CAD working stations to freeze for several minutes when searching for tests. The issue is here that the extension uses VS Code integrated search tool, which will try to match all files in the workspace against the pattern.

**Describe the solution you'd like**
I'm willing to contribute here but need project guidance and a direction decision. The following solutions come to my mind:

1. Check if the file pattern is actually not a pattern but an already fully qualified name. If that is the case skip the search and just take it (so setting or interface change).
2. Add an obvious alternative to `pattern` like `executable`.
3. Add an option to only search in the cwd and make the search pattern relative to that path.

**Describe alternatives you've considered**
N.A.

**Additional context**
- Windows 11
- VS Code 1.93.0
- C++ TestMate 4.12.0

**Comment by matepek** [comment]
Hello, 

Thanks for the issue and the PR although I'm not sure yet that which direction we should proceed. Let's discuss this topic a bit. So basically large build folders you have.

First thing came in to my mind, do you k ow about the `files.watcherExclude`?
See [createFileSystemWatcher API](https://code.visualstudio.com/api/references/vscode-api). Can't you just mark your build folder?  (See also: `search.exclude`)

The `vscode.workspace.findFiles` call which looks for the given pattern. THB I'm quite surprised that this scenario is not handed there. Would you be open to have a pair-debug session?

> Add an option to only search in the cwd and make the search pattern relative to that path.

What do you mean by this?

**Comment by armaxri** [comment]
Hi,

a pair debug session is a little bit difficult on the actual code base. But we could do it on a small example.

The hint with the `watcherExclude` was actually great!  By taking out Python venv´s and clangd´s index, as well as parts of the build, I was able to go down to a minute.

To give a better understanding of the issue I'll give a small overview of the setup, I'm trying to deliver.
My goal is to run unit and integration tests for a family of ventilators for ICUs. This is a highly regulated marked with very strict requirements which are forcing you to do many things by hand (instead of using libraries - you even build your own kernel to be sure, well two actually) and write many tests. The requirements and the amount of tests is so high, that actually Philips, GE and a few other companies just announced to leave the market.

Adding to the insane amount of files and tests, I need to list some of the test executables a few times due to product variants. So I need to test for small and big screen variants, as well as adults/kids and babies, which results in 4 runs of the same test but with different environment settings.

The extension is currently structured that it creates a watcher instance for each configured test (even if it is one executable, which is configured 4 times with different env variables). VS Code starts for each one `rg.exe`, which itself is actually a highly efficient tool. But with around 20 instances of the search tool, which is started on the workspace root directory, it will get quite extrem. As far as I understood the `rg.exe` tool, it checks recursively all files in the workspace (which are not ignored by the `files.watcherExclude`) and does a regex check against the pattern.

So without exclude I was seeing 20 searches on 300.000 files, which ended up in 6.000.000 regex checks.

With the exclude the computer can be used again after a minute and I would say that's ok'ish but still far from good. So my idea is to add the option to skip the search all along or to allow the user to limit the search area. For the limited search are one could use the `cwd` setting and start the search in that directory.

**Comment by matepek** [comment]
Can you show me your `settings.json` to see your `advancedExecutables` value?

**Comment by armaxri** [comment]
Sure
```.json
{
  "settings": {
    "testMate.cpp.test.advancedExecutables": [
      {
        "name": "Frontend Tests V600",
        "cwd": "${workspaceFolder}/build/Sim/install/bin",
        "pattern": "${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe",
        "runTask": {
          "before": ["prepare frontend tests V600"]
        },
        "env": {
          "PRODUCT": "V600"
        },
        "exclude": "files.watcherExclude"
      },
      {
        "name": "Frontend Tests V800",
        "cwd": "${workspaceFolder}/build/Sim/install/bin",
        "pattern": "${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe",
        "runTask": {
          "before": ["prepare frontend tests V800"]
        },
        "env": {
          "PRODUCT": "V800"
        },
        "exclude": "files.watcherExclude"
      },
      {
        "name": "Frontend Tests VN600",
        "cwd": "${workspaceFolder}/build/Sim/install/bin",
        "pattern": "${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe",
        "runTask": {
          "before": ["prepare frontend tests VN600"]
        },
        "env": {
          "PRODUCT": "VN600"
        },
        "exclude": "files.watcherExclude"
      },
      {
        "name": "Frontend Tests VN800",
        "cwd": "${workspaceFolder}/build/Sim/install/bin",
        "pattern": "${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe",
        "runTask": {
          "before": ["prepare frontend tests VN800"]
        },
        "env": {
          "PRODUCT": "VN800"
        },
        "exclude": "files.watcherExclude"
      }
    ]
  }
}
```

The env variable is actually a workaround. The extension doesn't recognise when there are different tasks that need to be executed beforehand.

**Comment by matepek** [comment]
How about the rest of the file?
Can you send me all?
Now I'm interested of `files.watcherExclude` too but might there are some details I find important.

**Comment by matepek** [comment]
> all locking for the same file

Which file is that?

How sure are you that test discovery mostly spends it's time waiting for `rg` to finish?

I could limit the parallel exploration. That might would decrease the UI impact but loading would still take a lot of time..

Might be related:
- https://github.com/Microsoft/vscode/issues/51948

**Comment by matepek** [comment]
Can you try [installing](https://stackoverflow.com/questions/42017617/how-to-install-vs-code-extension-manually) this attached version?

[vscode-catch2-test-adapter-4.12.2-beta.vsix.zip](https://github.com/user-attachments/files/17468692/vscode-catch2-test-adapter-4.12.2-beta.vsix.zip)

I'm very curious how the change detection work. I mean if you recompile your tests, does it detect and makes it a bit transparent? Did it before also?

**Comment by armaxri** [comment]
First of all: Your beta works great! The initial loading is way faster (even if I remove the `files.watcherExclude` settings)!

> > all locking for the same file
> 
> Which file is that?
In the example we have 4 search instances for the `${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe` file (if we configure it 4 times with different arguments. What I also was able to verify is that if you have 4 test cases `${workspaceFolder}/build/Sim/install/bin/FrontendTests.exe`  and lets say 2 `${workspaceFolder}/build/Sim/install/bin/BackendTests.exe`, you will get 6 processes which will look somehow like that (sorry I removed the tool for the actual monitoring but it really looks like that.
```
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/FrontendTests.exe
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/FrontendTests.exe
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/FrontendTests.exe
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/FrontendTests.exe
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/BackendTests.exe
<code_installation>/rg.exe --dir <workspace_root> --search <workspace_root>/build/Sim/install/bin/BackendTests.exe
```

With the 4.12.2-beta I don't see this structure anymore and actually have a good performance.

> How sure are you that test discovery mostly spends it's time waiting for `rg` to finish?
Very sure. I simply tracked the execution of the processes and monitored visually the end of the processes and the appearing of the tests.

> I could limit the parallel exploration. That might would decrease the UI impact but loading would still take a lot of time..
This is a road I would not recommend since it will kill the usage of the extension - can be extrem waiting times.

> Might be related:
> 
> * [what is rg.exe，it takes up much source microsoft/vscode#51948](https://github.com/microsoft/vscode/issues/51948)
I think so too.

So far the solution (I would like to know how you solved it at the end) looks fine. I just want to share my thought process and the issue with a simple example. The example shows a basic structure, which you will get when using CMake, a clean architecture and some bigger structures:
```.txt
project_repo
    - .cache_dir
        # clangd and other tools store quite a few files here.
    - build_dir
        - backend_dir
            - backend_lib.dll
        - frontend_dir
            - frontend_lib.dll
        - backend_test_dir
            # This dir contains a test, which cannot be executed due to missing DLLs.
            - backend_test_exe.exe
        - frontend_test_dir
            # This dir contains a test, which cannot be executed due to missing DLLs.
            - frontend_test_exe.exe
        - common_ui_dir
            - common_ui_lib.dll
        - install_dir
            # In a dir like this CMake collects all artifacts and here the executables can be executed.
            - common_ui_lib.dll
            - backend_lib.dll
            - frontend_lib.dll
            - backend_test_exe.exe
            - frontend_test_exe.exe
    - src_dir
        - backend_dir
            - backend_lib.cpp
        - frontend_dir
            - frontend_lib.cpp
    - test_dir
        - backend_test_dir
            - backend_test_exe.cpp
        - frontend_test_dir
            - frontend_test_exe.cpp
    - tools_dir
        - build_script.cmd
        - python_venv_dir
            # This content is heavy and needs to be excluded in the search.
    - shared_common_ui_repo_dir
        # This can be a git submodule or else. Devs can contribute but has it's own releases.
        - build_dir
            # This shared sub project runs it own tests for PRs and so on.
            # Therefore a custom local build is also required independent of the library consumer.
            - common_ui_dir
                - common_ui_lib.dll
            - common_ui_test_dir
                # This dir contains a test, which cannot be executed due to missing DLLs.
                - common_ui_test_exe.exe
            - install_dir
                # If the dev wants to contribute to this library, he wants to execute this as well.
                - common_ui_lib.dll
                - common_ui_test_exe.exe
        - src_dir
            - common_ui
                - common_ui_lib.cpp
        - test_dir
            - common_ui_test_dir
                - common_ui_test_exe.cpp
        - tools_dir
            - build_script.cmd
            - python_venv_dir
                # This content is heavy and needs to be excluded in the search.
```

When you take a look at the example, you will see that searching for executables does only make sense in two directories, despite the fact that there are other directories, which contain the test executables. Also setting the right filter can be quite a hassle. For the main `build_dir` you would need to make a very long exclude list for just one single directory, which should be included.

So thanks a lot for the new changes!

**Comment by matepek** [comment]
<details>
<summary>Fixed in <b>v4.12.2</b>.</summary>

This issue was mentioned in [CHANGELOG.md](./CHANGELOG.md) under a released entry so it is assumed to be fixed.
User verifications are always welcome.
</details>

**Comment by armaxri** [comment]
Thanks a lot! The solution looks very good and the performance so far is mind blowing!
